server/src/main/java/com/google/fhir/gateway/AuditEventHelperImpl.java

```diff
@@ -0,0 +1,425 @@
+      Reference agentUserWho,
+      Set<String> compartmentOwners) {
+
+    AuditEventBuilder auditEventBuilder =
+        initBaseAuditEventBuilder(requestDetailsReader, balpProfile, agentUserWho);
+
+    if (resource instanceof OperationOutcome) {
+
+      OperationOutcome.OperationOutcomeIssueComponent outcomeIssueComponent =
+          ((OperationOutcome) resource).getIssueFirstRep();
+
+      String errorDescription =
+          outcomeIssueComponent.getDetails().getText() != null
+              ? outcomeIssueComponent.getDetails().getText()
+              : outcomeIssueComponent.getDiagnostics();
+      auditEventBuilder.outcome(
+          AuditEventBuilder.Outcome.builder()
+              .code(mapOutcomeErrorCode(outcomeIssueComponent.getSeverity()))
+              .description(errorDescription)
+              .build());
+
+    } else {
+
+      if (compartmentOwners != null && !compartmentOwners.isEmpty()) {
+        for (String owner : compartmentOwners) {
+          auditEventBuilder.addEntityWhat(balpProfile, true, owner);
+        }
+      }
+
+      if (!ResourceType.Patient.equals(resource.getResourceType())) {
+        auditEventBuilder.addEntityWhat(balpProfile, false, extractLogicalId(resource));
+      }
+
+      auditEventBuilder.addQuery(requestDetailsReader);
+    }
+
+    return auditEventBuilder.build();
+  }
+
+  private AuditEvent createAuditEventCRUD(
+      RequestDetailsReader requestDetailsReader,
+      Resource resource,
+      BalpProfileEnum balpProfile,
+      Reference agentUserWho,
+      Set<String> compartmentOwners) {
+
+    AuditEventBuilder auditEventBuilder =
+        initBaseAuditEventBuilder(requestDetailsReader, balpProfile, agentUserWho);
+    if (resource instanceof OperationOutcome) {
+
+      OperationOutcome.OperationOutcomeIssueComponent outcomeIssueComponent =
+          ((OperationOutcome) resource).getIssueFirstRep();
+
+      String errorDescription =
+          outcomeIssueComponent.getDetails().getText() != null
+              ? outcomeIssueComponent.getDetails().getText()
+              : outcomeIssueComponent.getDiagnostics();
+      auditEventBuilder.outcome(
+          AuditEventBuilder.Outcome.builder()
+              .code(mapOutcomeErrorCode(outcomeIssueComponent.getSeverity()))
+              .description(errorDescription)
+              .build());
+
+    } else {
+      if (compartmentOwners != null && !compartmentOwners.isEmpty()) {
+        for (String owner : compartmentOwners) {
+          auditEventBuilder.addEntityWhat(balpProfile, true, owner);
+        }
+      }
+
+      if (!ResourceType.Patient.equals(resource.getResourceType())) {
+        auditEventBuilder.addEntityWhat(balpProfile, false, extractLogicalId(resource));
+      }
+    }
+
+    return auditEventBuilder.build();
+  }
+
+  private Reference createAgentClientWhoRef(RequestDetailsReader request) {
+    String clientId = JwtUtil.getClaimFromRequestDetails(request, JwtUtil.CLAIM_IHE_IUA_CLIENT_ID);
```

Comment:
As mentioned elsewhere, we should avoid decoding the JWT multiple times; it seems to me this is done several times when creating an AuditEvent.

Reply:
Done!